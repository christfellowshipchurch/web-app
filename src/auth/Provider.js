import React, {
    useReducer,
    useEffect,
    useState,
    useContext,
    createContext,
} from 'react';
import { AUTH_TOKEN_KEY } from '../keys';

const initialState = '';
const localState = localStorage.getItem(AUTH_TOKEN_KEY);
const AuthContext = createContext();
const useAuth = () => useContext(AuthContext);

const reducer = (token, newToken) => {
    if (newToken === null) {
        localStorage.removeItem(AUTH_TOKEN_KEY);
        return initialState;
    }
    return newToken || token;
};

const AuthProvider = (props) => {
    const [token, setToken] = useReducer(reducer, localState || initialState);
    const [isLoggedIn, setIsLoggedIn] = useState(!!token);
    const [triggerLogIn, setTriggerLogIn] = useState(false);
    const logout = () => {
        setToken(null);
    };

    useEffect(() => {
        localStorage.setItem(AUTH_TOKEN_KEY, token);
        setIsLoggedIn(!!token);
    }, [token]);

    return (
        <AuthContext.Provider
            value={{
                token,
                setToken,
                isLoggedIn,
                logout,
                triggerLogIn,
                logIn: () => {
                    if (!isLoggedIn) setTriggerLogIn(true);
                },
                hideLogIn: () => setTriggerLogIn(false),
            }}
        >
            {props.children}
        </AuthContext.Provider>
    );
};

export {
    AuthContext,
    AuthProvider,
    useAuth,
};
