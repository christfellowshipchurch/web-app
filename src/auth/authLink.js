import { setContext } from 'apollo-link-context';
import { AUTH_TOKEN_KEY } from '../keys';

export default setContext((_, { headers }) => {
  // get the authentication token from local storage if it exists
  const token = localStorage.getItem(AUTH_TOKEN_KEY);
  if (!token || token === '') {
    return {};
  }

  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      authorization: token || '',
    },
  };
});
