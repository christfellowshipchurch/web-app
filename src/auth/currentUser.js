import { useEffect } from 'react';
import { useQuery } from 'react-apollo';
import { useAuth } from '.';

const useAuthQuery = (query) => {
  const { token, isLoggedIn, logout } = useAuth();
  const { data, error, loading, refetch } = useQuery(query, {
    // if the user is not logged in,
    //  we want to skip the query all together
    //  and wait for refetch after the token changes
    skip: !isLoggedIn,
    // Required to ensure we always get fresh
    //  data from every request
    fetchPolicy: 'network-only',
    onError: () => {
      console.error('Authentication error: logging out');
      logout();
    },
  });

  useEffect(() => {
    // Catch statement is required if you want
    //  to manually handle the error callback
    //  specifically in regards to the refetch.
    // In this case, there's no need to make that
    //  error visible, so we're overwriting it with
    //  our own
    if (isLoggedIn) refetch().catch(() => console.error('Authentication error'));
  }, [token, isLoggedIn, refetch]);

  return {
    data,
    loading,
    error,
    isLoggedIn,
  };
};

export { useAuthQuery };
