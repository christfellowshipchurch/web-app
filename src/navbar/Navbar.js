import React, { useState } from 'react';
import { useQuery } from 'react-apollo';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { get, camelCase } from 'lodash';
import { Navbar } from 'react-bootstrap';

import { GET_WEBSITE_HEADER } from './queries';
import DynamicBanner from './DynamicBanner';
import DesktopNav from './DesktopNav';
import MobileNav from './MobileNav';

import NavbarToggle from './NavbarToggle';

// Takes a collection of images from the API's return data and formats
//  it to be an array of the following object structure: { imageKey: { uri, alt } }
const imageArrayToObject = (images) => {
  const imagesObj = {};

  images.forEach((n, i) => {
    const key = camelCase(get(n, 'name', i));
    const uri = get(n, 'sources[0].uri', '');
    const alt = get(n, 'name', 'Christ Fellowship Church');

    imagesObj[key] = { uri, alt };
  });

  return imagesObj;
};

const BrandImg = ({ className, uri, alt }) => (
  <Navbar.Brand href="/" className={classnames('align-self-start', className)}>
    <img src={uri} style={{ height: '58px', width: 'auto' }} alt={alt} />
  </Navbar.Brand>
);

const NavbarConnected = ({ bg, variant, brandImageKey, fixed, onToggle }) => {
  const [menuIcon, setMenuIcon] = useState(false);
  const website = process.env.REACT_APP_WEBSITE_KEY;
  const { loading, error, data } = useQuery(GET_WEBSITE_HEADER, {
    variables: { website },
    fetchPolicy: 'cache-and-network',
  });

  const navLinks = get(data, 'getWebsiteNavigation.navigationLinks', []);
  const quickAction = get(data, 'getWebsiteNavigation.quickAction', null);
  const menuLinks = get(data, 'getWebsiteNavigation.menuLinks', []);

  const images = imageArrayToObject(get(data, 'getWebsiteNavigation.images', []));
  const brandImage = get(images, brandImageKey, null);
  const navbarProps = {
    bg,
    variant,
    expand: 'lg',
  };

  const isDark = variant == 'dark';

  // We use sticky styling as the default so that padding is respected
  //    with the option to override it to use a fixed styling if preferred
  if (fixed) navbarProps.fixed = 'top';
  else navbarProps.sticky = 'top';

  return (
    <Navbar
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...navbarProps}
      id="user-profile-navbar-connected"
      className={classnames('d-flex', 'flex-column', 'p-0')}
    >
      <div className="row w-100">
        <DynamicBanner />
      </div>

      <div className={classnames('row', 'w-100', 'd-flex', 'justify-content-between')}>
        {/* { Right aligned Brand Image } */}
        {brandImage && (
          <BrandImg uri={brandImage.uri} alt={brandImage.alt} className="p-2 pl-3" />
        )}

        {/* { Toggle for Mobile } */}
        <NavbarToggle isDark={isDark} onClick={onToggle} />

        <Navbar.Collapse>
          <div
            className={classnames(
              'd-flex',
              'flex-column',

              'flex-lg-row',
              'justify-content-lg-end',

              'w-100',
              'px-lg-3'
            )}
          >
            {/* { Desktop Navigation } */}
            <DesktopNav
              navLinks={navLinks}
              menuLinks={menuLinks}
              quickAction={quickAction}
              isDark={isDark}
            />

            {/* { Mobile Navigation } */}
            <MobileNav
              navLinks={[...navLinks, ...menuLinks]}
              quickAction={quickAction}
              isDark={isDark}
            />
          </div>
        </Navbar.Collapse>
      </div>
    </Navbar>
  );
};

NavbarConnected.propTypes = {
  bg: PropTypes.string,
  variant: PropTypes.string,
  brandImageKey: PropTypes.string,
  fixed: PropTypes.bool,
  isDark: PropTypes.bool,
};

NavbarConnected.defaultProps = {
  bg: 'white',
  variant: 'light',
  brandImageKey: 'brandImage',
  fixed: false,
  isDark: false,
};

export default NavbarConnected;
