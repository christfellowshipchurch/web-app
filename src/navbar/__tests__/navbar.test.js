import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Navbar, { NavbarWithOpacity } from '../index';
import { AuthProvider } from '../../auth';

describe('Navbar', () => {
  it('renders without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <AuthProvider>
        <Navbar />
      </AuthProvider>
    );
    const tree = renderer.getRenderOutput();

    expect(tree).toMatchSnapshot();
  });
});

describe('NavbarWithOpacity', () => {
  it('renders without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <AuthProvider>
        <NavbarWithOpacity />
      </AuthProvider>
    );
    const tree = renderer.getRenderOutput();

    expect(tree).toMatchSnapshot();
  });
});
