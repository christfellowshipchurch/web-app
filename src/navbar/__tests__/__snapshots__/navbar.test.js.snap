// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders without crashing 1`] = `
<Context.Provider
  value={
    Object {
      "hideLogIn": [Function],
      "isLoggedIn": false,
      "logIn": [Function],
      "logout": [Function],
      "setToken": [Function],
      "token": "",
      "triggerLogIn": false,
    }
  }
>
  <NavbarConnected
    bg="white"
    brandImageKey="brandImage"
    fixed={false}
    onSelect={[Function]}
    onToggle={[Function]}
    variant="light"
  />
</Context.Provider>
`;
